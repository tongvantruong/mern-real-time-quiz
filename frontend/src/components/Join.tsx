import React, { memo, Profiler, useContext, useRef } from "react";
import { toast } from "react-toastify";
import { SocketContext } from "../context/socket";

type Props = {
  changeRoomId: (id: string) => void;
};

export const Join = memo(({ changeRoomId }: Props) => {
  const name = useRef<string>();
  const roomId = useRef<string>();

  const socket = useContext(SocketContext);

  function onFormSubmit(e: React.FormEvent<HTMLFormElement>) {
    e.preventDefault();
    const form = new FormData(e.currentTarget);
    const inputName = String(form.get("name"));

    if (!inputName || inputName.trim() === "") {
      toast("Please enter your name");
      return;
    }

    const inputRoomId = String(form.get("roomId"));
    if (!inputRoomId || inputRoomId.trim() === "") {
      toast("Please enter Quiz ID");
      return;
    }

    name.current = inputName;
    roomId.current = inputRoomId;
    changeRoomId(roomId.current);

    socket.emit("joinQuiz", roomId.current, name.current);
  }

  // to measure rendering performance of a React tree programmatically
  function onRender(
    id: unknown,
    phase: unknown,
    actualDuration: unknown,
    baseDuration: unknown,
    startTime: unknown,
    commitTime: unknown
  ) {
    console.log(`-------- render ${id} -------`);
    console.log(`phase: ${phase}`);
    console.log(`actualDuration: ${actualDuration}`);
    console.log(`baseDuration: ${baseDuration}`);
    console.log(`startTime: ${startTime}`);
    console.log(`commitTime: ${commitTime}`);
  }

  return (
    <Profiler id="Join" onRender={onRender}>
      <div className="flex flex-col gap-4 justify-center items-center mt-[-100px]">
        <h1>Join Real Time Quiz</h1>
        <form className="flex flex-col gap-4" onSubmit={(e) => onFormSubmit(e)}>
          <input
            className="rounded-lg text-violet-900 px-4 py-2 text-2xl bg-white focus:outline-none"
            type="text"
            name="name"
            placeholder="Enter your name"
          />
          <input
            name="roomId"
            className="rounded-lg text-violet-900 px-4 py-2 text-2xl bg-white focus:outline-none"
            type="text"
            placeholder="Enter a test ID"
          />
          <button
            className="bg-violet-600 hover:bg-violet-500 text-white mt-2"
            type="submit"
          >
            JOIN
          </button>
        </form>
      </div>
    </Profiler>
  );
});
