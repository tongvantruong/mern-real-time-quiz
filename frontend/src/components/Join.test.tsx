import { render, screen } from "@testing-library/react";
import { Join } from "./Join";
import { expect, it, vi } from "vitest";
import userEvent from "@testing-library/user-event";
import { SocketContext } from "../context/socket";
import io from "socket.io-client";

vi.mock("socket.io-client", () => {
  const mSocket = {
    emit: vi.fn(),
    on: vi.fn(),
    off: vi.fn(),
  };
  return { default: vi.fn(() => mSocket) };
});

it("should match snapshot", () => {
  const mockChangeRoomId = vi.fn();
  const { container } = render(
    <SocketContext.Provider value={io()}>
      <Join changeRoomId={mockChangeRoomId} />
    </SocketContext.Provider>
  );

  expect(container).toMatchSnapshot();
});

it("renders 2 input for name, quiz ID and a button", () => {
  const mockChangeRoomId = vi.fn();
  render(
    <SocketContext.Provider value={io()}>
      <Join changeRoomId={mockChangeRoomId} />
    </SocketContext.Provider>
  );

  expect(screen.getByPlaceholderText("Enter your name")).toBeInTheDocument();
  expect(screen.getByPlaceholderText("Enter a quiz ID")).toBeInTheDocument();
  expect(screen.getByText("JOIN")).toBeInTheDocument();
});

it("calls callback and emit joinQuiz when name, quiz id are provided", async () => {
  const mockChangeRoomId = vi.fn();
  render(
    <SocketContext.Provider value={io()}>
      <Join changeRoomId={mockChangeRoomId} />
    </SocketContext.Provider>
  );

  const name = "My name";
  const quizId = "1234";

  await userEvent.type(screen.getByPlaceholderText("Enter your name"), name);
  await userEvent.type(screen.getByPlaceholderText("Enter a quiz ID"), quizId);
  await userEvent.click(screen.getByText("JOIN"));

  expect(mockChangeRoomId).toHaveBeenCalledTimes(1);
  expect(mockChangeRoomId).toHaveBeenCalledWith(quizId);
  expect(io().emit).toHaveBeenCalledTimes(1);
  expect(io().emit).toHaveBeenCalledWith("joinQuiz", quizId, name);
});

it("does not call callback and emit joinQuiz when name is not provided", async () => {
  const mockChangeRoomId = vi.fn();
  render(
    <SocketContext.Provider value={io()}>
      <Join changeRoomId={mockChangeRoomId} />
    </SocketContext.Provider>
  );

  await userEvent.type(screen.getByPlaceholderText("Enter a quiz ID"), "123");
  await userEvent.click(screen.getByText("JOIN"));
  expect(mockChangeRoomId).toHaveBeenCalledTimes(0);
});

it("does not call callback and emit joinQuiz when quiz id is not provided", async () => {
  const mockChangeRoomId = vi.fn();
  render(
    <SocketContext.Provider value={io()}>
      <Join changeRoomId={mockChangeRoomId} />
    </SocketContext.Provider>
  );

  await userEvent.type(
    screen.getByPlaceholderText("Enter your name"),
    "My Name"
  );
  await userEvent.click(screen.getByText("JOIN"));
  expect(mockChangeRoomId).toHaveBeenCalledTimes(0);
});
