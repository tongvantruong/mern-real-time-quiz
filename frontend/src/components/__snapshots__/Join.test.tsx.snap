// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`should match snapshot 1`] = `
<div>
  <div
    class="flex flex-col gap-4 justify-center items-center mt-[-100px]"
  >
    <h1>
      Join Real Time Quiz
    </h1>
    <form
      class="flex flex-col gap-4"
    >
      <input
        class="rounded-lg text-violet-900 px-4 py-2 text-2xl bg-white focus:outline-none"
        name="name"
        placeholder="Enter your name"
        type="text"
      />
      <div
        class="relative"
      >
        <input
          class="rounded-lg text-violet-900 px-4 py-2 text-2xl bg-white focus:outline-none"
          name="roomId"
          placeholder="Enter a quiz ID"
          type="text"
        />
      </div>
      <button
        class="bg-violet-600 hover:bg-violet-500 text-white mt-2 outline-0"
        type="submit"
      >
        JOIN
      </button>
    </form>
  </div>
</div>
`;
